go/src/koding/klient/machine/machinegroup/clients/clients.go

```diff
@@ -0,0 +1,181 @@
+package clients
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"koding/klient/machine"
+
+	"github.com/koding/logging"
+)
+
+// ClientsOpts are the options used to configure set of dynamic clients.
+type ClientsOpts struct {
+	// Builder is a factory used to build clients.
+	Builder machine.ClientBuilder
+
+	// DynAddrInterval indicates how often dynamic client should pull address
+	// function looking for new addresses.
+	DynAddrInterval time.Duration
+
+	// PingInterval indicates how often dynamic client should ping external
+	// machine.
+	PingInterval time.Duration
+
+	// Log is used for logging. If nil, default logger will be created.
+	Log logging.Logger
+}
+
+// Valid checks if provided options are correct.
+func (opts *ClientsOpts) Valid() error {
+	if opts.Builder == nil {
+		return errors.New("nil client builder")
+	}
+	if opts.DynAddrInterval == 0 {
+		return errors.New("dynamic address check interval is not set")
+	}
+	if opts.PingInterval == 0 {
+		return errors.New("ping interval is not set")
+	}
+
+	return nil
+}
+
+// Clients is a set of dynamic clients binded to unique machine ID.
+type Clients struct {
+	builder         machine.ClientBuilder
+	dynAddrInterval time.Duration
+	pingInterval    time.Duration
+
+	log logging.Logger
+
+	mu sync.RWMutex
+	m  map[machine.ID]*machine.DynamicClient
+}
+
+// New creates a new Clients object.
+func New(opts *ClientsOpts) (*Clients, error) {
+	if err := opts.Valid(); err != nil {
+		return nil, err
+	}
+
+	c := &Clients{
+		builder:         opts.Builder,
+		dynAddrInterval: opts.DynAddrInterval,
+		pingInterval:    opts.PingInterval,
+		log:             opts.Log,
+		m:               make(map[machine.ID]*machine.DynamicClient),
+	}
+
+	if c.log == nil {
+		c.log = logging.NewLogger("")
+	}
+
+	return c, nil
+}
+
+// Create generates a new dynamic client for a given machine. If machine client
+// already exists, this function is no-op.
+func (c *Clients) Create(id machine.ID, dynAddr machine.DynamicAddrFunc) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.m[id]; ok {
+		return nil
+	}
+
+	dc, err := machine.NewDynamicClient(machine.DynamicClientOpts{
+		AddrFunc:        dynAddr,
+		Builder:         c.builder,
+		DynAddrInterval: c.dynAddrInterval,
+		PingInterval:    c.pingInterval,
+		Log:             c.log.New(string(id)),
+	})
+	if err != nil {
+		return err
+	}
+
+	c.m[id] = dc
+
+	return nil
+}
+
+// Drop closes and removes dynamic client binded to provided machine ID.
+func (c *Clients) Drop(id machine.ID) error {
+	c.mu.Lock()
+	dc, ok := c.m[id]
+	if !ok {
+		c.mu.Unlock()
+		return nil
+	}
+	delete(c.m, id)
+	c.mu.Unlock()
+
+	if dc != nil {
+		dc.Close()
+	} else {
+		c.log.Critical("Non existing client registered under %s ID", id)
```

Comment:
if it is for developers you should see the error while developing, so you can use panic, but if you say we can see this in the prod only, please add alerting to our logging system and if you get this message, fix the issue.